test/e2e/util/resource_manager.go

```diff
@@ -0,0 +1,316 @@
+package util
+
+import (
+	"context"
+	"fmt"
+	. "github.com/onsi/ginkgo"
+	"github.com/operator-framework/operator-lifecycle-manager/test/e2e/ctx"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	k8scontrollerclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+type K8sResourceManager interface {
+	Apply(obj k8scontrollerclient.Object, changeFunc interface{}) error
+	Create(obj k8scontrollerclient.Object, opts ...CreateOption) error
+	Delete(obj k8scontrollerclient.Object, opts ...DeleteOption) error
+	Get(obj k8scontrollerclient.Object, opts ...GetOption) error
+	GetByKey(namespace string, name string, obj k8scontrollerclient.Object, options ...GetOption) error
+	Update(obj k8scontrollerclient.Object) error
+	SetupTestNamespace(name string, options ...TestNamespaceOption) (*corev1.Namespace, error)
+	WaitForConditions(obj k8scontrollerclient.Object, conditions ...ObjectCondition) error
+	List(list k8scontrollerclient.ObjectList, options ...k8scontrollerclient.ListOption) error
+	WaitForEventWithReason(involvedObj k8scontrollerclient.Object, reason string) error
+	Reset(namespacesToDump ...string) error
+}
+
+func NewK8sResourceManager(ctx *ctx.TestContext) K8sResourceManager {
+	resourceManager := &k8sResourceManager{
+		testContext:      ctx,
+		createdResources: NewResourceQueue(),
+		backoff: wait.Backoff{
+			Duration: 1 * time.Second,
+			Jitter:   0.1,
+			Factor:   1.0,
+			Steps:    60,
+		},
+	}
+	_ = resourceManager.Reset()
+
+	return resourceManager
+}
+
+type k8sResourceManager struct {
+	testContext      *ctx.TestContext
+	kubeClient       k8scontrollerclient.Client
+	createdResources *ResourceQueue
+	backoff          wait.Backoff
+}
+
+func (m *k8sResourceManager) logf(str string, v ...interface{}) {
+	m.testContext.Logf(str, v...)
+}
+
+func (m *k8sResourceManager) Create(obj k8scontrollerclient.Object, opts ...CreateOption) error {
```

Comment:
A lot of the run-operation/test/retry methods have idiomatic and concise implementations using ginkgo/gomega. I think the concept of a resource manager makes a ton of sense for tracking, but I think all we should need to do is add a decorator to the dynamic controller-runtime client that tags resources for cleanup/removal based on the test running. In fact, it looks like this is very close to that already.